src/server/scripts/Commands/cs_learn.cpp

```diff
@@ -139,58 +139,50 @@ class learn_commandscript : public CommandScript
         return true;
     }
 
-    static bool HandleLearnAllMyClassCommand(ChatHandler* handler, char const* /*args*/)
+    static bool HandleLearnAllMyClassCommand(ChatHandler* handler)
     {
-        HandleLearnAllMySpellsCommand(handler, "");
-        HandleLearnAllMyTalentsCommand(handler, "");
+        HandleLearnAllMySpellsCommand(handler);
+        HandleLearnAllMyTalentsCommand(handler);
         return true;
     }
 
-    static bool HandleLearnAllMySpellsCommand(ChatHandler* handler, char const* /*args*/)
+    static bool HandleLearnAllMySpellsCommand(ChatHandler* handler)
     {
-        ChrClassesEntry const* classEntry = sChrClassesStore.LookupEntry(handler->GetSession()->GetPlayer()->GetClass());
+        ChrClassesEntry const* classEntry = sChrClassesStore.LookupEntry(handler->GetPlayer()->GetClass());
         if (!classEntry)
             return true;
-        uint32 family = classEntry->SpellClassSet;
 
-        for (uint32 i = 0; i < sSkillLineAbilityStore.GetNumRows(); ++i)
-        {
-            SkillLineAbilityEntry const* entry = sSkillLineAbilityStore.LookupEntry(i);
-            if (!entry)
-                continue;
-
-            SpellInfo const* spellInfo = sSpellMgr->GetSpellInfo(entry->Spell);
-            if (!spellInfo)
-                continue;
-
-            // skip server-side/triggered spells
-            if (spellInfo->SpellLevel == 0)
-                continue;
-
-            // skip wrong class/race skills
-            if (!handler->GetSession()->GetPlayer()->IsSpellFitByClassAndRace(spellInfo->Id))
-                continue;
-
-            // skip other spell families
-            if (spellInfo->SpellFamilyName != family)
-                continue;
+        Player* player = handler->GetPlayer();
+        auto trainers = sObjectMgr->GetClassTrainers(player->GetClass());
 
-            // skip spells with first rank learned as talent (and all talents then also)
-            if (GetTalentSpellCost(spellInfo->GetFirstRankSpell()->Id) > 0)
-                continue;
+        bool hadNew;
+        do
+        {
+            hadNew = false;
+            for (auto [_, trainer] : trainers)
```

Comment:
let me use your knowledge a little bit?
what means this [_ stuff?

Reply:
`trainers` is `unordered_multimap<uint8, Trainer const*>`

The `uint8` (class ID) binds to unused local `_`, the `Trainer const*` binds to `trainer`.

Reply:
oh... nice, ty <3